src/awkward/_namedaxis.py

```diff
@@ -0,0 +1,795 @@
+from __future__ import annotations
+
+import json
+import re
+from dataclasses import dataclass
+
+import awkward._typing as tp
+from awkward._regularize import is_integer
+
+# axis names are hashables, mostly strings,
+# except for integers, which are reserved for positional axis.
+AxisName: tp.TypeAlias = tp.Hashable
+
+# e.g.: {"x": 0, "y": 1, "z": 2}
+AxisMapping: tp.TypeAlias = tp.Mapping[AxisName, int]
+
+# e.g.: ("x", "y", None) where None is a wildcard
+AxisTuple: tp.TypeAlias = tp.Tuple[AxisName, ...]
+
+
+NAMED_AXIS_KEY: tp.Literal["__named_axis__"] = (
+    "__named_axis__"  # reserved for named axis
+)
+
+
+class AttrsNamedAxisMapping(tp.TypedDict, total=False):
+    __named_axis__: AxisMapping
+
+
+@tp.runtime_checkable
+class MaybeSupportsNamedAxis(tp.Protocol):
+    @property
+    def attrs(self) -> tp.Mapping | AttrsNamedAxisMapping: ...
+
+
+# just a class for inplace mutation
+class NamedAxis:
+    mapping: AxisMapping
+
+
+NamedAxis.mapping = {}
+
+
+def _prettify_named_axes(named_axis: AxisMapping) -> str:
+    """
+    Prettifies the named axes for better readability.
+
+    This function takes a named axis mapping and returns a string representation of the mapping.
+    The axis names are sorted in ascending order of their corresponding integer values.
+    If the axis name is a valid Python identifier, it is represented as is.
+    Otherwise, it is represented as a JSON string.
+
+    Args:
+        named_axis (AxisMapping): The named axis mapping to prettify.
+
+    Returns:
+        str: The prettified string representation of the named axis mapping.
+
+    Examples:
+        >>> _prettify_named_axes({"x": 0, "y": 1, "z": 2})
+        'x:0, y:1, z:2'
+        >>> _prettify_named_axes({"x": 0, "y": 1, "$": 2})
+        'x:0, y:1, "$":2'
+    """
+
+    def _prettify(ax: AxisName) -> str:
+        repr_ax = str(ax)
+        if re.match("[A-Za-z_][A-Za-z_0-9]*", repr_ax):
+            return repr_ax
+        return json.dumps(repr_ax)
+
+    sorted_named_axis = sorted(named_axis.items(), key=lambda x: x[1])
+    return ", ".join(
+        [f"{_prettify(named_ax)}:{pos_ax}" for named_ax, pos_ax in sorted_named_axis]
+    )
+
+
+def _neg2pos_axis(
+    axis: int,
+    total: int,
+) -> int:
+    """
+    Converts a negative axis index to a positive one.
+
+    This function takes a negative axis index and the total number of axes and returns the corresponding positive axis index.
+    If the input axis index is already positive, it is returned as is.
+
+    Args:
+        axis (int): The axis index to convert. Can be negative.
+        total (int): The total number of axes.
+
+    Returns:
+        int: The positive axis index corresponding to the input axis index.
+
+    Examples:
+        >>> _neg2pos_axis(-1, 3)
+        2
+        >>> _neg2pos_axis(1, 3)
+        1
+    """
+    if axis < 0:
+        return total + axis
+    return axis
+
```

Comment:
Is it worth connecting this with `maybe_posaxis`?

Reply:
This is a bit different to `maybe_posaxis` because it does not account for branching. Hm, but we could use `_neg2pos_axis` within `maybe_posaxis`... is this what you're suggesting?